ARCHITECTURE.md

```diff
@@ -0,0 +1,119 @@
+However, there are some other pieces of code useful during development which reside in the main
+directory:
+
+- `run-wheatley`: An executable Python script which will run the Wheatley code directly as though it
+  were run through `pip`.
+- `tests/*`: Unit tests for various pieces of Wheatley's code.  This only tests code in the
+  `wheatley/` folder.
+- `doctests`: Python script which invokes all the examples found in `README.md`, asserting that they
+  don't crash.  This prevents the examples from getting out of sync with the code.
+- `fuzz`, `fuzzing/*`: Fuzzing for CLI argument parsers.  These feed the parsing fuctions with
+  thousands of randomly generated inputs, asserting that they must produce well-defined errors.
+
+### `wheatley/{aliases.py, bell.py, calls.py, stroke.py}`
+
+These files contain little or no business logic, and instead provide datatypes (`bell.py`,
+`stroke.py`), type aliases (`aliases.py`) and/or constants that are used extensively throughout the
+code.  These serve two purposes: 
+1. They increase safety by providing an abstraction layer over the raw numbers and strings used to
+   communicate with Ringing Room.
+2. `bell.py` prevents any ambiguity between the many different representations of bell names (i.e.
+   is the 12th called `'T'` (name), `12` (number) or `11` (index)?) - Wheatley uses one `Bell` class
+   which provides unambiguous conversions to and from these representations.
+
+### `wheatley/bot.py`
+
+This is the glue code that holds Wheatley together.  The `Bot` class is a singleton that gets
+created at start-up and mediates interactions between other parts of the code (row generation,
+rhythm, interacting with Ringing Room, etc).  It also runs the `mainloop` - an infinite loop in
+which the main thread gets stuck until Wheatley stops.
+
+**Architectural Invariant**: None of the code in `row_generation/*`, `rhythm/*` or `tower.py` can
+talk directly to each other; instead they all provide an interface that the `Bot` class to mediate
+interactions.
+
+### `wheatley/rhythm/*`
+
+This is where the rhythm detection happens.  The specification of a rhythm is defined in
+`abstract_rhythm.py`, and there are two `Rhythm` classes which implement different behaviours:
+- `regression.py` uses regression to draw a linear line of best fit through the user's datapoints
+  and then uses this line to decide where Wheatley's bell should ring.
+- `wait_for_user.py` adds waiting for user-controlled bells on top of an existing `Rhythm` class.
+
+**Architectural Invariant**: The rhythm module should never access the real time directly - it
+should use the times passed to each individual method.  This is because `WaitForUserRhythm` lies to
+its internal rhythm in order to stop Wheatley from jumping back to the original rhythm if someone
+holds up for a long time.
+
+### `wheatley/row_generation/*`
+
+This is the code that determines _what_ Wheatley rings, and reacts to `Bob`s and `Single`s.
```

Comment:
```suggestion
This is the code that determines _what_ Wheatley rings, and reacts to `Bob` and `Single`.
```
The trailing s looks a bit odd when rendered

Reply:
I've tweaked the rest of the sentence to make it easier to read, but this has been fixed.